Multiple cache invalidation not working with a native build

### Describe the bug

Hello, 
I have a REST endpoint to invalidate all my caches like so:

```
    @DELETE
    @CacheInvalidateAll(cacheName = Caches.MY_CACHE_A)
    @CacheInvalidateAll(cacheName = Caches.MY_CACHE_B)
    @CacheInvalidateAll(cacheName = Caches.MY_CACHE_C)
    public void invalidateCaches() {
        logger.info("Invalidating declared caches");
    }
```

When running in dev mode, everything is fine. The DEBUG logs shows that all caches are properly invalidated :

```
2021-11-03 15:06:12,757 DEBUG [io.qua.cac.run.CacheInvalidateAllInterceptor] (executor-thread-0) Invalidating all entries from cache ["cacheA"]
2021-11-03 15:06:12,759 DEBUG [io.qua.cac.run.CacheInvalidateAllInterceptor] (executor-thread-0) Invalidating all entries from cache ["cacheB"]
2021-11-03 15:06:12,759 DEBUG [io.qua.cac.run.CacheInvalidateAllInterceptor] (executor-thread-0) Invalidating all entries from cache ["cacheC"]
```

When running in native mode, a bug investigation revealed that only one cache is invalidated. This was reproduced on locahost. Calls to other resources use data from the cache instead of calling the backends.
`2021-11-03 15:06:12,759 DEBUG [io.qua.cac.run.CacheInvalidateAllInterceptor] (executor-thread-0) Invalidating all entries from cache ["cacheB"]`

(i have changed the cache names above but even after changing the order of the 3 annotations, it was always the same cache that was invalidated)


A temporary workaround is to have my multi-annotated method to delegate the invalidation to as many sub methods as necessary, each one with its own invalidation. Obviously that is the opposite of the interest to have multiple annotations. 

(And as a side note, i know this is somewhere on the roadmap: being able to programmatically invalidate all caches from my enum instead of adding/forgetting to add an annotation would even be better)

Thanks for your help !
Pascal.


### Expected behavior

All caches mentionned in each annotation are flushed.

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

Linux Sunnyvale 5.4.0-89-generic #100-Ubuntu SMP Fri Sep 24 14:50:10 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "11.0.12" 2021-07-20

### GraalVM version (if different from Java)

OpenJDK Runtime Environment GraalVM CE 21.2.0 (build 11.0.12+6-jvmci-21.2-b08)

### Quarkus version or git rev

2.1.4.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 6.8.3

### Additional information

_No response_

Here's a reproducer: https://github.com/gwenneg/quarkus-cache-native

The JVM tests are OK but the native version is failing:
```
Executing "/home/glepage/git/quarkus-cache-native/target/code-with-quarkus-1.0.0-SNAPSHOT-runner -Dquarkus.http.port=8081 -Dquarkus.http.ssl-port=8444 -Dtest.url=http://localhost:8081 -Dquarkus.log.file.path=/home/glepage/git/quarkus-cache-native/target/quarkus.log -Dquarkus.log.file.enable=true"
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2021-11-05 10:34:12,915 INFO  [io.quarkus] (main) code-with-quarkus 1.0.0-SNAPSHOT native (powered by Quarkus 2.4.1.Final) started in 0.011s. Listening on: http://0.0.0.0:8081
2021-11-05 10:34:12,915 INFO  [io.quarkus] (main) Profile prod activated. 
2021-11-05 10:34:12,915 INFO  [io.quarkus] (main) Installed features: [cache, cdi, resteasy, resteasy-jackson, smallrye-context-propagation, vertx]
[ERROR] WARNING: An illegal reflective access operation has occurred
[ERROR] WARNING: Illegal reflective access by org.codehaus.groovy.vmplugin.v9.Java9 (file:/home/glepage/.m2/repository/org/codehaus/groovy/groovy/3.0.8/groovy-3.0.8.jar) to constructor java.lang.AssertionError(java.lang.String)
[ERROR] WARNING: Please consider reporting this to the maintainers of org.codehaus.groovy.vmplugin.v9.Java9
[ERROR] WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
[ERROR] WARNING: All illegal access operations will be denied in a future release
2021-11-05 10:34:13,898 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-0) HTTP Request to /test-cache/method1 failed, error id: 628fa176-d87a-4c4e-91df-22fcd94cb402-1: org.jboss.resteasy.spi.UnhandledException: java.lang.IllegalArgumentException: duplicate element: @io.quarkus.cache.CacheInvalidateAll()
	at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:381)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:218)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:519)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:135)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:90)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:543)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:829)
	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:567)
	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
Caused by: java.lang.IllegalArgumentException: duplicate element: @io.quarkus.cache.CacheInvalidateAll()
	at java.util.ImmutableCollections$Set12.<init>(ImmutableCollections.java:510)
	at java.util.Set.of(Set.java:699)
	at org.acme.GreetingResource_Subclass.<init>(GreetingResource_Subclass.zig:163)
	at org.acme.GreetingResource_Bean.create(GreetingResource_Bean.zig:183)
	at org.acme.GreetingResource_Bean.create(GreetingResource_Bean.zig:199)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:96)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:29)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:26)
	at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:26)
	at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:26)
	at io.quarkus.arc.impl.ClientProxies.getApplicationScopedDelegate(ClientProxies.java:17)
	at org.acme.GreetingResource_ClientProxy.arc$delegate(GreetingResource_ClientProxy.zig:67)
	at org.acme.GreetingResource_ClientProxy.arc_contextualInstance(GreetingResource_ClientProxy.zig:82)
	at io.quarkus.arc.runtime.ClientProxyUnwrapper.apply(ClientProxyUnwrapper.java:11)
	at io.quarkus.resteasy.common.runtime.QuarkusInjectorFactory$UnwrappingPropertyInjector.inject(QuarkusInjectorFactory.java:74)
	at org.jboss.resteasy.plugins.server.resourcefactory.POJOResourceFactory.createResource(POJOResourceFactory.java:80)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:401)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:69)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
	... 17 more

[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 1.716 s <<< FAILURE! - in org.acme.NativeGreetingResourceIT
[ERROR] org.acme.NativeGreetingResourceIT.testCache  Time elapsed: 0.931 s  <<< FAILURE!
java.lang.AssertionError: 
1 expectation failed.
Expected status code <200> but was <500>.
```

I had the same issue with my initial quick test. That's when I decided to add log entries in quarkus-cache and check what was happening. These entries showed me that `InterceptorBindings#getInterceptorBindings` returns null in native mode.

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ArcProcessor.java
@@ -469,6 +469,7 @@ public class ArcProcessor {
             public void registerField(FieldInfo fieldInfo) {
                 reflectiveFields.produce(new ReflectiveFieldBuildItem(fieldInfo));
             }
+
         }, existingClasses.existingClasses, bytecodeTransformerConsumer,
                 config.shouldEnableBeanRemoval() && config.detectUnusedFalsePositives);
         for (ResourceOutput.Resource resource : resources) {
@@ -495,6 +496,11 @@ public class ArcProcessor {
         // Register all qualifiers for reflection to support type-safe resolution at runtime in native image
         for (ClassInfo qualifier : beanProcessor.getBeanDeployment().getQualifiers()) {
             reflectiveClasses.produce(new ReflectiveClassBuildItem(true, false, qualifier.name().toString()));
+        }
+
+        // Register all interceptor bindings for reflection so that AnnotationLiteral.equals() works in a native image
+        for (ClassInfo binding : beanProcessor.getBeanDeployment().getInterceptorBindings()) {
+            reflectiveClasses.produce(new ReflectiveClassBuildItem(true, false, binding.name().toString()));
         }
 
         ArcContainer container = recorder.getContainer(shutdown);

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/AbstractGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/AbstractGenerator.java
@@ -14,9 +14,15 @@ abstract class AbstractGenerator {
     static final String SYNTHETIC_SUFFIX = "Synthetic";
 
     protected final boolean generateSources;
+    protected final ReflectionRegistration reflectionRegistration;
+
+    public AbstractGenerator(boolean generateSources, ReflectionRegistration reflectionRegistration) {
+        this.generateSources = generateSources;
+        this.reflectionRegistration = reflectionRegistration;
+    }
 
     public AbstractGenerator(boolean generateSources) {
-        this.generateSources = generateSources;
+        this(generateSources, null);
     }
 
     /**

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/AnnotationLiteralGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/AnnotationLiteralGenerator.java
@@ -48,11 +48,8 @@ public class AnnotationLiteralGenerator extends AbstractGenerator {
 
     private static final Logger LOGGER = Logger.getLogger(AnnotationLiteralGenerator.class);
 
-    private final boolean generateSources;
-
     AnnotationLiteralGenerator(boolean generateSources) {
         super(generateSources);
-        this.generateSources = generateSources;
     }
 
     /**
@@ -72,7 +69,7 @@ public class AnnotationLiteralGenerator extends AbstractGenerator {
         return resources;
     }
 
-    static void createSharedAnnotationLiteral(ClassOutput classOutput, Key key, Literal literal, Set<String> existingClasses) {
+    void createSharedAnnotationLiteral(ClassOutput classOutput, Key key, Literal literal, Set<String> existingClasses) {
         // Ljavax/enterprise/util/AnnotationLiteral<Lcom/foo/MyQualifier;>;Lcom/foo/MyQualifier;
         String signature = String.format("Ljavax/enterprise/util/AnnotationLiteral<L%1$s;>;L%1$s;",
                 key.annotationName.toString().replace('.', '/'));
@@ -113,6 +110,7 @@ public class AnnotationLiteralGenerator extends AbstractGenerator {
         generateStaticFieldsWithDefaultValues(annotationLiteral, defaultOfClassType);
 
         annotationLiteral.close();
+
         LOGGER.debugf("Shared annotation literal generated: %s", literal.className);
     }
 

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanGenerator.java
@@ -86,7 +86,6 @@ public class BeanGenerator extends AbstractGenerator {
     protected final AnnotationLiteralProcessor annotationLiterals;
     protected final Predicate<DotName> applicationClassPredicate;
     protected final PrivateMembersCollector privateMembers;
-    protected final ReflectionRegistration reflectionRegistration;
     protected final Set<String> existingClasses;
     protected final Map<BeanInfo, String> beanToGeneratedName;
     protected final Predicate<DotName> injectionPointAnnotationsPredicate;
@@ -97,11 +96,10 @@ public class BeanGenerator extends AbstractGenerator {
             Set<String> existingClasses, Map<BeanInfo, String> beanToGeneratedName,
             Predicate<DotName> injectionPointAnnotationsPredicate,
             List<Function<BeanInfo, Consumer<BytecodeCreator>>> suppressConditionGenerators) {
-        super(generateSources);
+        super(generateSources, reflectionRegistration);
         this.annotationLiterals = annotationLiterals;
         this.applicationClassPredicate = applicationClassPredicate;
         this.privateMembers = privateMembers;
-        this.reflectionRegistration = reflectionRegistration;
         this.existingClasses = existingClasses;
         this.beanToGeneratedName = beanToGeneratedName;
         this.injectionPointAnnotationsPredicate = injectionPointAnnotationsPredicate;
@@ -222,13 +220,11 @@ public class BeanGenerator extends AbstractGenerator {
         implementGetIdentifier(bean, beanCreator);
         implementSupplierGet(beanCreator);
         if (!bean.hasDefaultDestroy()) {
-            implementDestroy(bean, beanCreator, providerType, Collections.emptyMap(), reflectionRegistration,
-                    isApplicationClass, baseName);
+            implementDestroy(bean, beanCreator, providerType, Collections.emptyMap(), isApplicationClass, baseName);
         }
         implementCreate(classOutput, beanCreator, bean, providerType, baseName,
                 Collections.emptyMap(), Collections.emptyMap(),
-                Collections.emptyMap(), reflectionRegistration,
-                targetPackage, isApplicationClass);
+                Collections.emptyMap(), targetPackage, isApplicationClass);
         implementGet(bean, beanCreator, providerType, baseName);
 
         implementGetTypes(beanCreator, beanTypes.getFieldDescriptor());
@@ -312,15 +308,14 @@ public class BeanGenerator extends AbstractGenerator {
         implementGetIdentifier(bean, beanCreator);
         implementSupplierGet(beanCreator);
         if (!bean.hasDefaultDestroy()) {
-            implementDestroy(bean, beanCreator, providerType, injectionPointToProviderSupplierField,
-                    reflectionRegistration,
-                    isApplicationClass, baseName);
+            implementDestroy(bean, beanCreator, providerType, injectionPointToProviderSupplierField, isApplicationClass,
+                    baseName);
         }
         implementCreate(classOutput, beanCreator, bean, providerType, baseName,
                 injectionPointToProviderSupplierField,
                 interceptorToProviderSupplierField,
                 decoratorToProviderSupplierField,
-                reflectionRegistration, targetPackage, isApplicationClass);
+                targetPackage, isApplicationClass);
         implementGet(bean, beanCreator, providerType, baseName);
 
         implementGetTypes(beanCreator, beanTypes.getFieldDescriptor());
@@ -416,13 +411,12 @@ public class BeanGenerator extends AbstractGenerator {
         implementGetIdentifier(bean, beanCreator);
         implementSupplierGet(beanCreator);
         if (!bean.hasDefaultDestroy()) {
-            implementDestroy(bean, beanCreator, providerType, injectionPointToProviderField, reflectionRegistration,
-                    isApplicationClass, baseName);
+            implementDestroy(bean, beanCreator, providerType, injectionPointToProviderField, isApplicationClass, baseName);
         }
         implementCreate(classOutput, beanCreator, bean, providerType, baseName,
                 injectionPointToProviderField,
                 Collections.emptyMap(), Collections.emptyMap(),
-                reflectionRegistration, targetPackage, isApplicationClass);
+                targetPackage, isApplicationClass);
         implementGet(bean, beanCreator, providerType, baseName);
 
         implementGetTypes(beanCreator, beanTypes.getFieldDescriptor());
@@ -505,13 +499,11 @@ public class BeanGenerator extends AbstractGenerator {
         implementGetIdentifier(bean, beanCreator);
         implementSupplierGet(beanCreator);
         if (!bean.hasDefaultDestroy()) {
-            implementDestroy(bean, beanCreator, providerType, null, reflectionRegistration, isApplicationClass,
-                    baseName);
+            implementDestroy(bean, beanCreator, providerType, null, isApplicationClass, baseName);
         }
         implementCreate(classOutput, beanCreator, bean, providerType, baseName,
                 Collections.emptyMap(), Collections.emptyMap(),
-                Collections.emptyMap(), reflectionRegistration,
-                targetPackage, isApplicationClass);
+                Collections.emptyMap(), targetPackage, isApplicationClass);
         implementGet(bean, beanCreator, providerType, baseName);
 
         implementGetTypes(beanCreator, beanTypes.getFieldDescriptor());
@@ -727,7 +719,7 @@ public class BeanGenerator extends AbstractGenerator {
         constructor.writeInstanceField(
                 FieldDescriptor.of(beanCreator.getClassName(), FIELD_NAME_BEAN_TYPES, Set.class.getName()),
                 constructor.getThis(),
-                constructor.invokeStaticInterfaceMethod(MethodDescriptors.SET_OF,
+                constructor.invokeStaticMethod(MethodDescriptors.SETS_OF,
                         typesArray));
 
         // Qualifiers
@@ -750,7 +742,7 @@ public class BeanGenerator extends AbstractGenerator {
             constructor.writeInstanceField(
                     FieldDescriptor.of(beanCreator.getClassName(), FIELD_NAME_QUALIFIERS, Set.class.getName()),
                     constructor.getThis(),
-                    constructor.invokeStaticInterfaceMethod(MethodDescriptors.SET_OF,
+                    constructor.invokeStaticMethod(MethodDescriptors.SETS_OF,
                             qualifiersArray));
         }
 
@@ -765,15 +757,14 @@ public class BeanGenerator extends AbstractGenerator {
             constructor.writeInstanceField(
                     FieldDescriptor.of(beanCreator.getClassName(), FIELD_NAME_STEREOTYPES, Set.class.getName()),
                     constructor.getThis(),
-                    constructor.invokeStaticInterfaceMethod(MethodDescriptors.SET_OF,
+                    constructor.invokeStaticMethod(MethodDescriptors.SETS_OF,
                             stereotypesArray));
         }
         return constructor;
     }
 
     protected void implementDestroy(BeanInfo bean, ClassCreator beanCreator, ProviderType providerType,
-            Map<InjectionPointInfo, String> injectionPointToProviderField, ReflectionRegistration reflectionRegistration,
-            boolean isApplicationClass, String baseName) {
+            Map<InjectionPointInfo, String> injectionPointToProviderField, boolean isApplicationClass, String baseName) {
 
         MethodCreator destroy = beanCreator
                 .getMethodCreator("destroy", void.class, providerType.descriptorName(), CreationalContext.class)
@@ -908,7 +899,7 @@ public class BeanGenerator extends AbstractGenerator {
             Map<InjectionPointInfo, String> injectionPointToProviderSupplierField,
             Map<InterceptorInfo, String> interceptorToProviderSupplierField,
             Map<DecoratorInfo, String> decoratorToProviderSupplierField,
-            ReflectionRegistration reflectionRegistration, String targetPackage, boolean isApplicationClass) {
+            String targetPackage, boolean isApplicationClass) {
 
         MethodCreator create = beanCreator.getMethodCreator("create", providerType.descriptorName(), CreationalContext.class)
                 .setModifiers(ACC_PUBLIC);
@@ -1410,7 +1401,7 @@ public class BeanGenerator extends AbstractGenerator {
             ResultHandle invocationContextHandle = create.invokeStaticMethod(
                     MethodDescriptors.INVOCATION_CONTEXTS_AROUND_CONSTRUCT, constructorHandle,
                     aroundConstructsHandle, func.getInstance(),
-                    create.invokeStaticInterfaceMethod(MethodDescriptors.SET_OF, bindingsArray));
+                    create.invokeStaticMethod(MethodDescriptors.SETS_OF, bindingsArray));
             TryBlock tryCatch = create.tryBlock();
             CatchBlockCreator exceptionCatch = tryCatch.addCatch(Exception.class);
             // throw new RuntimeException(e)
@@ -1542,7 +1533,7 @@ public class BeanGenerator extends AbstractGenerator {
             // InvocationContextImpl.postConstruct(instance,postConstructs).proceed()
             ResultHandle invocationContextHandle = create.invokeStaticMethod(
                     MethodDescriptors.INVOCATION_CONTEXTS_POST_CONSTRUCT, instanceHandle,
-                    postConstructsHandle, create.invokeStaticInterfaceMethod(MethodDescriptors.SET_OF, bindingsArray));
+                    postConstructsHandle, create.invokeStaticMethod(MethodDescriptors.SETS_OF, bindingsArray));
 
             TryBlock tryCatch = create.tryBlock();
             CatchBlockCreator exceptionCatch = tryCatch.addCatch(Exception.class);

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ClientProxyGenerator.java
@@ -57,15 +57,13 @@ public class ClientProxyGenerator extends AbstractGenerator {
 
     private final Predicate<DotName> applicationClassPredicate;
     private final boolean mockable;
-    private final ReflectionRegistration reflectionRegistration;
     private final Set<String> existingClasses;
 
     public ClientProxyGenerator(Predicate<DotName> applicationClassPredicate, boolean generateSources, boolean mockable,
             ReflectionRegistration reflectionRegistration, Set<String> existingClasses) {
-        super(generateSources);
+        super(generateSources, reflectionRegistration);
         this.applicationClassPredicate = applicationClassPredicate;
         this.mockable = mockable;
-        this.reflectionRegistration = reflectionRegistration;
         this.existingClasses = existingClasses;
     }
 

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/DecoratorGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/DecoratorGenerator.java
@@ -106,9 +106,8 @@ public class DecoratorGenerator extends BeanGenerator {
         implementGetIdentifier(decorator, decoratorCreator);
         implementSupplierGet(decoratorCreator);
         implementCreate(classOutput, decoratorCreator, decorator, providerType, baseName,
-                injectionPointToProviderField,
-                Collections.emptyMap(), Collections.emptyMap(),
-                reflectionRegistration, targetPackage, isApplicationClass);
+                injectionPointToProviderField, Collections.emptyMap(), Collections.emptyMap(),
+                targetPackage, isApplicationClass);
         implementGet(decorator, decoratorCreator, providerType, baseName);
         implementGetTypes(decoratorCreator, beanTypes.getFieldDescriptor());
         implementGetBeanClass(decorator, decoratorCreator);

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/InterceptorGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/InterceptorGenerator.java
@@ -101,7 +101,7 @@ public class InterceptorGenerator extends BeanGenerator {
         implementCreate(classOutput, interceptorCreator, interceptor, providerType, baseName,
                 injectionPointToProviderField,
                 Collections.emptyMap(), Collections.emptyMap(),
-                reflectionRegistration, targetPackage, isApplicationClass);
+                targetPackage, isApplicationClass);
         implementGet(interceptor, interceptorCreator, providerType, baseName);
         implementGetTypes(interceptorCreator, beanTypes.getFieldDescriptor());
         implementGetBeanClass(interceptor, interceptorCreator);

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/MethodDescriptors.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/MethodDescriptors.java
@@ -20,6 +20,7 @@ import io.quarkus.arc.impl.MapValueSupplier;
 import io.quarkus.arc.impl.Objects;
 import io.quarkus.arc.impl.Reflections;
 import io.quarkus.arc.impl.RemovedBeanImpl;
+import io.quarkus.arc.impl.Sets;
 import io.quarkus.gizmo.MethodDescriptor;
 import java.lang.reflect.Constructor;
 import java.lang.reflect.Field;
@@ -223,7 +224,7 @@ public final class MethodDescriptors {
     public static final MethodDescriptor COLLECTIONS_EMPTY_MAP = MethodDescriptor.ofMethod(Collections.class, "emptyMap",
             Map.class);
 
-    public static final MethodDescriptor SET_OF = MethodDescriptor.ofMethod(Set.class, "of", Set.class, Object[].class);
+    public static final MethodDescriptor SETS_OF = MethodDescriptor.ofMethod(Sets.class, "of", Set.class, Object[].class);
 
     public static final MethodDescriptor ARC_CONTAINER = MethodDescriptor.ofMethod(Arc.class, "container", ArcContainer.class);
 

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/ObserverGenerator.java
@@ -64,7 +64,6 @@ public class ObserverGenerator extends AbstractGenerator {
     private final AnnotationLiteralProcessor annotationLiterals;
     private final Predicate<DotName> applicationClassPredicate;
     private final PrivateMembersCollector privateMembers;
-    private final ReflectionRegistration reflectionRegistration;
     private final Set<String> existingClasses;
     private final Map<ObserverInfo, String> observerToGeneratedName;
     private final Predicate<DotName> injectionPointAnnotationsPredicate;
@@ -74,11 +73,10 @@ public class ObserverGenerator extends AbstractGenerator {
             PrivateMembersCollector privateMembers, boolean generateSources, ReflectionRegistration reflectionRegistration,
             Set<String> existingClasses, Map<ObserverInfo, String> observerToGeneratedName,
             Predicate<DotName> injectionPointAnnotationsPredicate, boolean mockable) {
-        super(generateSources);
+        super(generateSources, reflectionRegistration);
         this.annotationLiterals = annotationLiterals;
         this.applicationClassPredicate = applicationClassPredicate;
         this.privateMembers = privateMembers;
-        this.reflectionRegistration = reflectionRegistration;
         this.existingClasses = existingClasses;
         this.observerToGeneratedName = observerToGeneratedName;
         this.injectionPointAnnotationsPredicate = injectionPointAnnotationsPredicate;
@@ -567,7 +565,7 @@ public class ObserverGenerator extends AbstractGenerator {
             constructor.writeInstanceField(
                     FieldDescriptor.of(observerCreator.getClassName(), "qualifiers", Set.class.getName()),
                     constructor.getThis(),
-                    constructor.invokeStaticInterfaceMethod(MethodDescriptors.SET_OF,
+                    constructor.invokeStaticMethod(MethodDescriptors.SETS_OF,
                             qualifiersArray));
         }
 

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java
@@ -77,7 +77,6 @@ public class SubclassGenerator extends AbstractGenerator {
             "bindings", Set.class);
 
     private final Predicate<DotName> applicationClassPredicate;
-    private final ReflectionRegistration reflectionRegistration;
     private final Set<String> existingClasses;
 
     static String generatedName(DotName providerTypeName, String baseName) {
@@ -90,10 +89,9 @@ public class SubclassGenerator extends AbstractGenerator {
     public SubclassGenerator(AnnotationLiteralProcessor annotationLiterals, Predicate<DotName> applicationClassPredicate,
             boolean generateSources, ReflectionRegistration reflectionRegistration,
             Set<String> existingClasses) {
-        super(generateSources);
+        super(generateSources, reflectionRegistration);
         this.applicationClassPredicate = applicationClassPredicate;
         this.annotationLiterals = annotationLiterals;
-        this.reflectionRegistration = reflectionRegistration;
         this.existingClasses = existingClasses;
     }
 
@@ -259,7 +257,7 @@ public class SubclassGenerator extends AbstractGenerator {
                         constructor.writeArrayValue(bindingsArray, bindingsIndex++,
                                 bindingsLiterals.computeIfAbsent(binding, bindingsLiteralFun));
                     }
-                    return constructor.invokeStaticInterfaceMethod(MethodDescriptors.SET_OF, bindingsArray);
+                    return constructor.invokeStaticMethod(MethodDescriptors.SETS_OF, bindingsArray);
                 }
             }
         };
@@ -817,7 +815,7 @@ public class SubclassGenerator extends AbstractGenerator {
             // InvocationContextImpl.preDestroy(this,predestroys)
             ResultHandle invocationContext = tryCatch.invokeStaticMethod(MethodDescriptors.INVOCATION_CONTEXTS_PRE_DESTROY,
                     tryCatch.getThis(), predestroysHandle,
-                    tryCatch.invokeStaticInterfaceMethod(MethodDescriptors.SET_OF, bindingsArray));
+                    tryCatch.invokeStaticMethod(MethodDescriptors.SETS_OF, bindingsArray));
 
             // InvocationContext.proceed()
             tryCatch.invokeInterfaceMethod(MethodDescriptors.INVOCATION_CONTEXT_PROCEED, invocationContext);

--- a/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Sets.java
+++ b/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/Sets.java
@@ -1,0 +1,33 @@
+package io.quarkus.arc.impl;
+
+import java.util.Arrays;
+import java.util.Set;
+
+public final class Sets {
+
+    private Sets() {
+    }
+
+    /**
+     * Unlike {@link Set#of(Object...)} this method does not throw an {@link IllegalArgumentException} if there are duplicate
+     * elements.
+     * 
+     * @param <E>
+     * @param elements
+     * @return the set
+     */
+    @SafeVarargs
+    public static <E> Set<E> of(E... elements) {
+        switch (elements.length) {
+            case 0:
+                return Set.of();
+            case 1:
+                return Set.of(elements[0]);
+            case 2:
+                return elements[0].equals(elements[1]) ? Set.of(elements[0]) : Set.of(elements[0], elements[1]);
+            default:
+                return Set.copyOf(Arrays.asList(elements));
+        }
+    }
+
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/arc/interceptor/Simple.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/arc/interceptor/Simple.java
@@ -1,0 +1,33 @@
+package io.quarkus.it.arc.interceptor;
+
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.ElementType.TYPE;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Repeatable;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import javax.enterprise.util.Nonbinding;
+import javax.interceptor.InterceptorBinding;
+
+@Repeatable(Simple.List.class)
+@Target({ TYPE, METHOD })
+@Retention(RUNTIME)
+@InterceptorBinding
+public @interface Simple {
+
+    @Nonbinding
+    String name();
+
+    @Retention(RetentionPolicy.RUNTIME)
+    @Target(ElementType.METHOD)
+    @interface List {
+
+        Simple[] value();
+
+    }
+
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/arc/interceptor/SimpleBean.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/arc/interceptor/SimpleBean.java
@@ -1,0 +1,14 @@
+package io.quarkus.it.arc.interceptor;
+
+import javax.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped
+public class SimpleBean {
+
+    @Simple(name = "foo")
+    @Simple(name = "bar")
+    public String ping() {
+        return "OK";
+    }
+
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/arc/interceptor/SimpleInterceptor.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/arc/interceptor/SimpleInterceptor.java
@@ -1,0 +1,17 @@
+package io.quarkus.it.arc.interceptor;
+
+import javax.annotation.Priority;
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.Interceptor;
+import javax.interceptor.InvocationContext;
+
+@Simple(name = "")
+@Priority(1)
+@Interceptor
+public class SimpleInterceptor {
+
+    @AroundInvoke
+    Object mySuperCoolAroundInvoke(InvocationContext ctx) throws Exception {
+        return ctx.proceed();
+    }
+}

--- a/integration-tests/main/src/main/java/io/quarkus/it/arc/interceptor/TestSimpleBeanEndpoint.java
+++ b/integration-tests/main/src/main/java/io/quarkus/it/arc/interceptor/TestSimpleBeanEndpoint.java
@@ -1,0 +1,19 @@
+package io.quarkus.it.arc.interceptor;
+
+import javax.inject.Inject;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+
+@Path("/simple-bean")
+public class TestSimpleBeanEndpoint {
+
+    @Inject
+    SimpleBean simpleBean;
+
+    @GET
+    public String manualValidation() {
+        return simpleBean.ping();
+
+    }
+
+}

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/SimpleBeanITCase.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/SimpleBeanITCase.java
@@ -1,0 +1,8 @@
+package io.quarkus.it.main;
+
+import io.quarkus.test.junit.QuarkusIntegrationTest;
+
+@QuarkusIntegrationTest
+public class SimpleBeanITCase extends SimpleBeanTestCase {
+
+}

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/SimpleBeanTestCase.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/SimpleBeanTestCase.java
@@ -1,0 +1,19 @@
+package io.quarkus.it.main;
+
+import static org.hamcrest.Matchers.is;
+
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.test.junit.QuarkusTest;
+import io.restassured.RestAssured;
+
+@QuarkusTest
+public class SimpleBeanTestCase {
+
+    @Test
+    public void testRequestScope() {
+        RestAssured.when().get("/simple-bean").then()
+                .body(is("OK"));
+    }
+
+}

